Migration from/to symlink/rsync deployment and .rsync-filter for shared folders (rsync deployment)

mblaschke opened the pull request:
Prevent deletion of shared folders while rsync to current/ and minimize downtime

ricardclau commented:
:-1: here for me, IMHO this should be added in the repo to deploy and rsync will use it

ricardclau commented:
Oh sorry, that's the feature you added recently, let me check it out :)

mblaschke commented:
Will add a commit later..please wait for merge :)

mblaschke commented:
This feature minimizes downtime by preveting the deletion of shared folders while rsync is running for "rsync current deployment" and offers also the possibility for a rollback feature without knowing which folders/files are shared folders.

Eg. if there is a main configuration file (local configuration with database connection) the rsync task will delete it and it takes some seconds until ansible restores it.

The rsync-filter prevent the deletion of these shared folders and minimze the downtime of the application.

I've moved the creation of `.rsync-filter` from `rsync-deploy.yml` to `symlink-shared.yml` task because it's customizable inside the hooks (`ansistrano_after_symlink_tasks_file` or `ansistrano_before_cleanup_tasks_file`) after.

Also i've added a migration from symlink to rsync deployment and back.

mblaschke commented:
@ricardclau This feature is finished, please check again :)

ricardclau commented on tasks/rsync-deploy.yml:
```diff
@@ -15,13 +27,6 @@
-    path: "{{ ansistrano_deploy_to }}/{{ ansistrano_current_dir }}/{{ item }}"
-  with_items: "{{ ansistrano_shared_paths }}"
-
 # Symlinks shared paths
```
That bit should be removed as well right? Please check #115 (although this was merged in your previous PR)

mblaschke replied:
That would reduce downtime during deployment (first all symlinks are removed, than all symlinks are readded).

Would be better to remove and add each symlink to shared one after one.

I've reverted that change because if shared folders are changing (eg. a existing folder is now a shared folder) the deployment failes because ansible cannot create the symlink.
I'm still trying to find a good solution.

ricardclau commented on tasks/rsync-deploy.yml:
```diff
@@ -1,4 +1,16 @@
+- stat:
+    path: "{{ ansistrano_deploy_to }}/{{ ansistrano_current_dir }}"
+  register: stat_current_dir
+
```
Like this bit!

mblaschke replied:
See #115 for a explanations why there is a special symlink handling is needed for rsync deployment.

ricardclau commented:
It would be great if you could add some test using your bit

mblaschke commented:
Still working on this branch, currently i've found a issue with `synchronize` task because it's using an ssh loop back connection on the server and failes some times (maybe bacause of known hosts issues).

I would go back to `command: rsync ...` which will fix this issue.

mblaschke commented:
@ricardclau can you check again? Are these tests enough?
We're currently testing it as deployment system with also automatically switching from symlink to directory. Since switching from synchronize to local rsync without any trouble.

ricardclau commented on tasks/rsync-deploy.yml:
```diff
@@ -1,19 +1,24 @@
-    delete: yes
-    archive: yes
-  delegate_to: "{{ inventory_hostname }}"
+  command: rsync -a -F --delete-after "{{ ansistrano_release_path.stdout }}/" "{{ ansistrano_deploy_to }}/{{ ansistrano_current_dir }}/"
```
why not using synchronize here?

mblaschke replied:
Because it's using ssh (remote box --> ssh to remote box --> rsync) for a simple rsync command.
This can be an issue for some servers and isn't very fast - too much overhead for this simple task.

ricardclau replied:
There is not much overhead for this so I would prefer to stay with synchronize as it uses all the right flags and it is more descriptive

mblaschke replied:
But you have to deal with `.ssh/known_hosts` and outgoing connections and maybe DNS issues when using rsync via ssh that's why i've replaced it with a plain rsync. It's also much faster here then running via SSH.

The rsync strategy is not the best method for deployment but a symlink is not always possible (eg. shared hosts) so my goal was to keep the trouble and downtime at a minimum. For some beginners that can be really an issue to setup a working environment.
Also there are companies that doesn't allow these connections because of security issues and they don't want to change the policy.

Eg. if you have a two digit number of servers with multiple accounts then you have to setup a valid `.ssh/known_hosts` for each account on each server for automatic deployment just because a loopback connection of one single task which can be run locally.

ricardclau replied:
Wow, does the local rsync use known_hosts when using synchronize? That's bonkers!

Alright, will merge this tonight it as it does not affect the "main" Ansistrano path and you seem to need this :)

How does rollback end up in this scenario? I am not sure it will work as it relies on symlinks, have you tested it? Anyway, if it is broken it was broken since we merged your first iteration

mblaschke replied:
It starts a ssh connection from the remote server to the remote server (loopback) so the known_hosts file on the remote server will be used.
Maybe i find a way to prevent the usage of the ssh connection but currently it's a big problem :(

The rollback would be just a rsync from a previous release but there is still one issue: the shared folders must be reset if they changed between deployments. If they don't change the `.rsync-filter` will protect the shared folders.
But this functionality still be must be implemented into rollback.

ricardclau replied:
I see, I think there is some parameter to avoid this known_hosts file situation but I can´t remember and we can always sort it out later

I need to do some tests with the .rsync-filter as it has been a while since I don´t dig into all the options we have :)
Do you want to give a go to rollback? Or at least review what we would do there?

ricardclau commented on tasks/symlink-shared.yml:
```diff
@@ -13,3 +13,19 @@
     src: "{{ item | regex_replace('[^\\/]*', '..') }}/../shared/{{ item }}"
   with_items: "{{ ansistrano_shared_paths }}"
+
+# Remove previous .rsync-filter file (rsync current deployment)
```
Not sure about this, what happens if the deployment strategy is rsync and there is a .rsync-filter file in your repo (like I do for instance sometimes)

mblaschke replied:
Then the .rsync-filter file is removed before setup the new one.

mblaschke commented:
Found another small issue.. will fix it in some minutes :)

ricardclau commented:
Cool, will wait for your fix then

mblaschke commented:
I think i've fixed all issues (was just the rsync time issue with non root user).
Our deployment is working with root and non root users with different servers.

ricardclau commented:
Cool, thanks for all your work and constant discussion here. Nice stuff!

Will do some more testing this weekend with no root user and merge it soon
